Support dynamic node IP changes

zshi-redhat opened the pull request:
Adjust ovnk configurations based on primary IP change:
- OVNK related node annotation
- OVN encap IP
- Egress IP

Signed-off-by: Zenghui Shi <zshi@redhat.com>
Co-authored-by: Andreas Karis <ak.karis@gmail.com>

trozet commented on go-controller/pkg/node/node_ip_handler_linux.go:
```diff
@@ -25,17 +26,22 @@ type addressManager struct {
 
 // initializes a new address manager which will hold all the IPs on a node
-func newAddressManager(nodeName string, k kube.Interface, config *managementPortConfig, watchFactory factory.NodeWatchFactory) *addressManager {
+func newAddressManager(nodeName string, k kube.Interface, config *managementPortConfig, watchFactory factory.NodeWatchFactory, gwBridge *bridgeConfiguration) *addressManager {
```
I'm worried this isn't thread safe. The gwBridge *bridgeConfiguration is also used by the openflow manager thread. wdyt?

zshi-redhat replied:
@trozet good point! In current implementation, it may be safe that the `gwBridge.bridgeName` in AddressManager is used for retrieving the primary interface IP addresses and updating node annotation, we didn't modify any fields (e.g. ips) in gwBridge that are used by openflow manager. In case it does need to modify gwBridge in AddressManager in the future, I think we can add a function to update the openflowManager struct data in the nodeIPManager.onChanged callback. Does it sound reasonable or maybe you have other considerations?

trozet replied:
I'm thinking in the future other threads may also read/write from *bridgeConfiguration. Should we just add a mutex to that struct now to protect it?

zshi-redhat replied:
Added the sync.Mutex in bridgeConfiguration struct.

zshi-redhat commented:
/retest

trozet commented on go-controller/pkg/node/node_ip_handler_linux.go:
```diff
@@ -181,6 +185,57 @@ func (c *addressManager) Run(stopChan <-chan struct{}, doneWg *sync.WaitGroup) {
+	}
+
+	// get updated interface IP addresses for the gateway bridge
+	ifAddrs, err := getNetworkInterfaceIPAddresses(c.gatewayBridge.bridgeName)
```
dont you need to lock the bridge here while you read from it? Also lock the bridge in the other thread?

zshi-redhat replied:
yes, working on it.

zshi-redhat replied:
done

zshi-redhat commented:
/retest

zshi-redhat commented:
/retest

andreaskaris commented on go-controller/pkg/node/node_ip_handler_linux.go:
```diff
@@ -196,11 +196,15 @@ func (c *addressManager) updateNodeAddressAnnotations() {
 	}
 
 	// get updated interface IP addresses for the gateway bridge
+	c.gatewayBridge.Lock()
```
Move into a helper function and call `defer c.gatewayBridge.Lock()`? May be overkill, but my preference would be unlocking in a single location and ideally in a defer
~~~
ifAddrs, err := c.getNetworkInterfaceIPAddresses()
if err != nil {
    klog.Errorf("Failed to get gateway bridge IP addresses: %v", err)
         return
    }
}  
~~~

~~~
// getNetworkInterfaceIPAddresses sets and returns the bridge's current ifAddresses.
func (c *addressManager) getNetworkInterfaceIPAddresses() ([]*net.IPNet, error) {
  c.gatewayBridge.Lock()
  defer c.gatewayBridge.Unlock()
  ifAddrs, err := getNetworkInterfaceIPAddresses(c.gatewayBridge.bridgeName)
  if err == nil {
    c.gatewayBridge.ips = ifAddrs
  } 
 return ifAddrs, err
}
~~~

zshi-redhat replied:
@andreaskaris Thanks for the review! applied the suggestions.

andreaskaris commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -1591,7 +1612,7 @@ func newSharedGateway(nodeName string, subnets []*net.IPNet, gwNextHops []net.IP
 			}
 			klog.Info("Creating Shared Gateway Node Port Watcher")
-			gw.nodePortWatcher, err = newNodePortWatcher(gwBridge.patchPort, gwBridge.bridgeName, gwBridge.uplinkName, nodeName, gwBridge.ips, gw.openflowManager, gw.nodeIPManager, watchFactory)
+			gw.nodePortWatcher, err = newNodePortWatcher(gwBridge, nodeName, gw.openflowManager, gw.nodeIPManager, watchFactory)
```
nice!

andreaskaris commented:
I am currently rewriting the E2E tests that I had originally written for this and this PR is missing a node watcher. 

The tunnel encap IP information comes from the API's node object's status.

Therefore it works when:
* you update kubelet configuration and restart kubelet first (updates the node object), then update the IP address (because that's when the AddressManager's IP address watcher gets triggered and reads from the node object)

It does not work when:
* you update the IP address first (that's when the AddressManager's IP address watcher gets triggered) and then update the kubelet configuration and restart it (because nothing gets triggered but the node object is updated)

I'll continue working on this tomorrow but we need to use the node watcher that's part of the address manager and add some update logic there to fix this

zshi-redhat commented on go-controller/pkg/node/gateway_shared_intf.go:
```diff
@@ -524,6 +527,18 @@ func serviceUpdateNotNeeded(old, new *kapi.Service) bool {
 		!util.EgressSVCHostChanged(old, new)
 }
 
+func updateNodePortWatcher(nodePortWatcher *nodePortWatcher, addressManager *addressManager) {
```
Applied the suggestions and resolved merge conflict.

zshi-redhat commented:
> I am currently rewriting the E2E tests that I had originally written for this and this PR is missing a node watcher.
Cool, nice!
> 
> The tunnel encap IP information comes from the API's node object's status.
> 
> Therefore it works when:
> 
>     * you update kubelet configuration and restart kubelet first (updates the node object), then update the IP address (because that's when the AddressManager's IP address watcher gets triggered and reads from the node object)
> 
> 
> It does not work when:
> 
>     * you update the IP address first (that's when the AddressManager's IP address watcher gets triggered) and then update the kubelet configuration and restart it (because nothing gets triggered but the node object is updated)
> 

Do we expect ovnkube-node pod to restart when kubelet restarts? 
The node logic to update encap IP is in the setupOVNNode path, maybe it won't be reconciled if ovnkube-node doesn't restart?
> 
> I'll continue working on this tomorrow but we need to use the node watcher that's part of the address manager and add some update logic there to fix this; I'll prob open a PR against your branch once I'm done so that you can pull it in? I should be able to invest a good amount of time here this week

Sure, sounds good. I updated the PR to take your suggestions in.

zshi-redhat commented:
/retest

andreaskaris commented:
I'm just copy/pasting this from the slack channel to here, for completeness
================

As part of https://github.com/ovn-org/ovn-kubernetes/pull/3131 (dynamic node IP changes) we update the OVS encap IP when
we detect an IP address change on a node.
We simply run:
{code}
ovs-vsctl set Open_vSwitch . external_ids:ovn-encap-ip=<new IP>
{code}

Whereas this fixes outbound traffic immediately, we see for ca. 3 minutes that the remote hosts do not update
their endpoints, yet.

Trigger update from 172.18.0.3 to 172.18.255.254.

ovn-worker (my session to the pod closed, the update occurred even earlier than 14:43:01, but good enough for illustration purposes):
{code}
sh-5.2# while true; do sleep 5; date; ovs-vsctl list open . | grep encap; done
Tue Dec 13 14:43:01 UTC 2022
external_ids        : {hostname=ovn-worker, ovn-bridge-mappings="physnet:breth0", ovn-enable-lflow-cache="true", ovn-encap-ip="172.18.255.254", ovn-encap-type=geneve, ovn-monitor-all="true", ovn-ofctrl-wait-before-clear="0", ovn-openflow-probe-interval="180", ovn-remote="tcp:172.18.0.4:6642", ovn-remote-probe-interval="100000", rundir="/var/run/openvswitch", system-id="bb108a75-097b-4c17-851e-f4e3276e295a"}
(...)
Tue Dec 13 14:43:57 UTC 2022
external_ids        : {hostname=ovn-worker, ovn-bridge-mappings="physnet:breth0", ovn-enable-lflow-cache="true", ovn-encap-ip="172.18.255.254", ovn-encap-type=geneve, ovn-monitor-all="true", ovn-ofctrl-wait-before-clear="0", ovn-openflow-probe-interval="180", ovn-remote="tcp:172.18.0.4:6642", ovn-remote-probe-interval="100000", rundir="/var/run/openvswitch", system-id="bb108a75-097b-4c17-851e-f4e3276e295a"
(...)
Tue Dec 13 14:44:50 UTC 2022
external_ids        : {hostname=ovn-worker, ovn-bridge-mappings="physnet:breth0", ovn-enable-lflow-cache="true", ovn-encap-ip="172.18.255.254", ovn-encap-type=geneve, ovn-monitor-all="true", ovn-ofctrl-wait-before-clear="0", ovn-openflow-probe-interval="180", ovn-remote="tcp:172.18.0.4:6642", ovn-remote-probe-interval="100000", rundir="/var/run/openvswitch", system-id="bb108a75-097b-4c17-851e-f4e3276e295a"}
(...)
{code}

ovn-worker2:
{code}
sh-5.2# while true; do sleep 5; date; ovs-vsctl show | grep remote_ip; done
(...)
Tue Dec 13 14:41:41 UTC 2022
                options: {csum="true", key=flow, remote_ip="172.18.0.3"}
                options: {csum="true", key=flow, remote_ip="172.18.0.4"}
(...)
Tue Dec 13 14:44:55 UTC 2022
                options: {csum="true", key=flow, remote_ip="172.18.0.3"}
                options: {csum="true", key=flow, remote_ip="172.18.0.4"}
Tue Dec 13 14:45:00 UTC 2022
                options: {csum="true", key=flow, remote_ip="172.18.0.4"}
                options: {csum="true", key=flow, remote_ip="172.18.255.254"}
{code}

andreaskaris commented:
Dumitru helped me figure this out. The ovn-controller running on the node that changes IP doesn't reconnect to the SB until 200 seconds later:
~~~
2022-12-14T13:48:47Z|01882|reconnect|DBG|tcp:172.18.255.254:49868: connecting...
~~~

At that point it will download the contents of the SB and create its new chassis and encap records with the new IP.
~~~
ovnkube-node-n8nsw-ovn-controller.log:2022-12-14T13:48:46.087Z|64277|reconnect|ERR|tcp:172.18.0.2:6642: no response to inactivity probe after 100 seconds, disconnecting
~~~

The inactivity probe is 100 seconds and it will only detect that the SB connection is gone due to the inactivity probe.
Which causes the connection to tear down after 2 * probe-interval without a reply, so 200 seconds. That explains the entire delay then and why I have to add at least a 180 seconds sleep.

We could play around with those delays and make them shorter if we wanted to go down that route --- or we have to restart ovn-controller (which I think is more logical in our case).

If we wanted to lower the delays, this was Ilya's suggestion to LD_PRELOAD the keepalive library https://mail.openvswitch.org/pipermail/ovs-dev/2022-August/396878.html
So the CMS (CNO/ovn-kube/whatever) could do this when starting ovn-controller. We'd have to do it for the whole lifetime of the process.

andreaskaris commented:
@zshi-redhat  https://github.com/zshi-redhat/ovn-kubernetes/pull/1

zshi-redhat commented:
> 
> It does not work when:
> 
> * you update the IP address first (that's when the AddressManager's IP address watcher gets triggered) and then update the kubelet configuration and restart it (because nothing gets triggered but the node object is updated)
> 
@andreaskaris  The encap-ip is updated by AddressManager by subscribing to the netlink change, why is it related to the node object if we don't restart ovnkube-node?

zshi-redhat commented:
> If we wanted to lower the delays, this was Ilya's suggestion to LD_PRELOAD the keepalive library https://mail.openvswitch.org/pipermail/ovs-dev/2022-August/396878.html So the CMS (CNO/ovn-kube/whatever) could do this when starting ovn-controller. We'd have to do it for the whole lifetime of the process.

my understanding is we don't handle the ovn-controller restart in this PR or ovn-kubernetes. The CMS (CNO) will need to detect the node IP change and restart ovn-controller or ovnkube-node ds, right?

andreaskaris commented:
> The encap-ip is updated by AddressManager by subscribing to the netlink change, why is it related to the node object if we don't restart ovnkube-node?

The AddressManager watches changes to the link (interface). So when you update br-ex to add a new IP address, it will be triggered. It then retrieves the node.Status.InternalIP from the API node object. This works when you:
first update kubelet.conf and restart kubelet. the node.status will change here. Then you change the interface IP address, the addressManager is triggered due to the change on the interface, and it can fetch the updated node.status

This does not work when you:
first update the interface IP. The addressManager is triggered. It looks into the node.Status which still contains kubelets old configuration, the old IP address. No IP address change is detected, the encap IP is not updated. You now modify kubelet config and restart kubelet. This will udpate the node.status. But the addressManager is not run again, because it only reacts to IP address changes on the interface.

If you look into my E2E test, I cover both possible orders of operation, and if you run the E2E tests for the second case against your original PR, they fail.

Conclusion: the addressManager also must monitor changes to the node object, otherwise it will not work for the second use case

zshi-redhat commented:
> > > 
> > 
> > 
> > my understanding is we don't handle the ovn-controller restart in this PR or ovn-kubernetes. The CMS (CNO) will need to detect the node IP change and restart ovn-controller or ovnkube-node ds, right?
> 
> Yep, I agree. If a few minutes of downtime are o.k., then ovn-controller does not even have to be restarted. My E2Es against upstream ovn-kubernetes work without restarting ovn-controller with a 200 second sleep to account for the relatively long OVN probes - after 200 seconds, ovn-controller reconnects from the new source IP.

ack+, thanks for adding the e2e tests! I felt pleasant reviewing the tests!
> 
> I'd say a follow-up PR to CNO (probably) is needed to either change the .yaml file for ovn-kube so that it monitors for IP address changes and restarts the pod or some other mechanism to restart ovn-controller (or the entire pod); well or we make it part of the procedure so the admin has to delete the ovnkube-node pod

If it ends up with restarting ovnkube-node pod, then we probably don't need node IP change detection in ovnkube-node since it will re-initialize the encap-ip, node annotation etc during startup, right?  So the fix could be that the CMS detects the IP change in node object and restart ovnkube-node pod (both ovn-controller and ovnkube-node).

zshi-redhat commented:
/retest

andreaskaris commented:
> If it ends up with restarting ovnkube-node pod, then we probably don't need node IP change detection in ovnkube-node since it will re-initialize the encap-ip, node annotation etc during startup, right? So the fix could be that the CMS detects the IP change in node object and restart ovnkube-node pod (both ovn-controller and ovnkube-node).

Funny, right? That's what I thought, too, yesterday when writing my answer. Like how much of the work on this commit would then not be needed and we only need to be able to detect things on ovnkube-node startup and adjust them accordingly ... ;-)

zshi-redhat commented:
> > If it ends up with restarting ovnkube-node pod, then we probably don't need node IP change detection in ovnkube-node since it will re-initialize the encap-ip, node annotation etc during startup, right? So the fix could be that the CMS detects the IP change in node object and restart ovnkube-node pod (both ovn-controller and ovnkube-node).
> 
> Funny, right? That's what I thought, too, yesterday when writing my answer. Like how much of the work on this commit would then not be needed and we only need to detect an IP address change, kill the pod, and then the ovnkube code only needs to be able to detect things on ovnkube-node startup and adjust them accordingly ...

maybe it is useful when moving to ovn-ic, which allows ovn-controller to reconnect sb via local sock?

zshi-redhat commented:
> > > If it ends up with restarting ovnkube-node pod, then we probably don't need node IP change detection in ovnkube-node since it will re-initialize the encap-ip, node annotation etc during startup, right? So the fix could be that the CMS detects the IP change in node object and restart ovnkube-node pod (both ovn-controller and ovnkube-node).
> > 
> > 
> > Funny, right? That's what I thought, too, yesterday when writing my answer. Like how much of the work on this commit would then not be needed and we only need to detect an IP address change, kill the pod, and then the ovnkube code only needs to be able to detect things on ovnkube-node startup and adjust them accordingly ...
> 
> maybe it is useful when moving to ovn-ic, which allows ovn-controller to reconnect sb via local sock?

Add Dumitru's comment here:

When sock is used, ovn-controller will be able to talk to SB regardless of IP change. But the problem is that if somehow the SB goes down ovn-controller might not be able to determine on its own that the connection went away (for example if ovn-controller doesn't need to send data to the SB which is often the case).   Although for the sock case we might not want to restart ovn-controller every time the SB goes down.  Maybe we should have a way to tell ovn-controller to reconnect.

zshi-redhat commented:
@trozet PTAL.

zshi-redhat commented:
> > > > If it ends up with restarting ovnkube-node pod, then we probably don't need node IP change detection in ovnkube-node since it will re-initialize the encap-ip, node annotation etc during startup, right? So the fix could be that the CMS detects the IP change in node object and restart ovnkube-node pod (both ovn-controller and ovnkube-node).
> > > 
> > > 
> > > Funny, right? That's what I thought, too, yesterday when writing my answer. Like how much of the work on this commit would then not be needed and we only need to detect an IP address change, kill the pod, and then the ovnkube code only needs to be able to detect things on ovnkube-node startup and adjust them accordingly ...
> > 
> > 
> > maybe it is useful when moving to ovn-ic, which allows ovn-controller to reconnect sb via local sock?
> 
> Add Dumitru's comment here:
> 
> When sock is used, ovn-controller will be able to talk to SB regardless of IP change. But the problem is that if somehow the SB goes down ovn-controller might not be able to determine on its own that the connection went away (for example if ovn-controller doesn't need to send data to the SB which is often the case). Although for the sock case we might not want to restart ovn-controller every time the SB goes down. Maybe we should have a way to tell ovn-controller to reconnect.

The ovn-controller reconnect is not a problem with unix connections, see https://bugzilla.redhat.com/show_bug.cgi?id=2154796.
With ovn-ic using unix connection, we don't need the CMS workaround to restart ovn-controller, this PR is still needed.

andreaskaris commented:
@trozet just a reminder

trozet commented:
@andreaskaris @zshi-redhat  why dont we just exit ovn-controller if using tcp and ip changes via "ovn-appctl -t ovn-controller exit". ovnkube-node can execute this command.

trozet commented on test/e2e/node_ip_migration.go:
```diff
@@ -0,0 +1,680 @@
+				})
+
+				for _, updateKubeletFirst := range []bool{true, false} {
+					updateKubeletFirst := updateKubeletFirst // Required to avoid race conditions due to pointer assignment.
```
not seeing why this is needed

zshi-redhat replied:
@andreaskaris  could you help on Tim's question?

trozet commented on test/e2e/node_ip_migration.go:
```diff
@@ -0,0 +1,680 @@
+				}
+			})
+
+			Context("when EgressIPs are configured", func() {
```
what is the relationship between egress IP and node ip address migration?

zshi-redhat replied:
noreroute LR policies will be updated when node IP changes.

trozet commented:
please squash the fixup commits too

zshi-redhat commented on go-controller/pkg/node/node_ip_handler_linux.go:
```diff
@@ -201,16 +256,61 @@ func (c *addressManager) doesNodeHostAddressesMatch() bool {
+func (c *addressManager) nodePrimaryAddrChanged() bool {
+	node, err := c.watchFactory.GetNode(c.nodeName)
+	if err != nil {
+		klog.Errorf("Unable to get node from informer")
```
fixed to use caller to handle the logging message.

zshi-redhat commented on go-controller/pkg/node/node_ip_handler_linux.go:
```diff
@@ -201,16 +256,61 @@ func (c *addressManager) doesNodeHostAddressesMatch() bool {
 
+// nodePrimaryAddrChanged returns false if the node's primary API IP,
+// i.e. the first NodeInternalIP or NodeExternalIP from node.Status.Addresses
+// does not match the current address stored in c.nodePrimaryAddr.
```
updated the function comment and fixed the logging.

zshi-redhat commented on go-controller/pkg/node/node_ip_handler_linux.go:
```diff
@@ -181,6 +185,57 @@ func (c *addressManager) Run(stopChan <-chan struct{}, doneWg *sync.WaitGroup) {
+	if err := c.nodeAnnotator.Run(); err != nil {
+		klog.Errorf("Failed to set node annotations, err: %v", err)
+	}
+	c.OnChanged()
```
moved the c.OnChanged() to its original place which will be executed regardless of annotation failure.

zshi-redhat commented on go-controller/pkg/node/node_ip_handler_linux.go:
```diff
@@ -181,6 +185,57 @@ func (c *addressManager) Run(stopChan <-chan struct{}, doneWg *sync.WaitGroup) {
+	// Get node information
+	node, err := c.watchFactory.GetNode(c.nodeName)
+	if err != nil {
+		klog.Errorf("Unable to get node from informer: %v", err)
```
updated.

zshi-redhat commented on go-controller/pkg/node/gateway.go:
```diff
@@ -360,6 +360,7 @@ func (g *gateway) GetGatewayBridgeIface() string {
 }
 
 type bridgeConfiguration struct {
```
EgressIP was missing in last version, added it back.
upon node IP change, no-reroute logical router policy (102) in ovn_cluster_router will be updated.

zshi-redhat commented:
> please squash the fixup commits too

done.

trozet commented:
@zshi-redhat what about this comment https://github.com/ovn-org/ovn-kubernetes/pull/3131#issuecomment-1378026374 ?

zshi-redhat commented:
> @andreaskaris @zshi-redhat why dont we just exit ovn-controller if using tcp and ip changes via "ovn-appctl -t ovn-controller exit". ovnkube-node can execute this command.

@trozet sorry I missed this comment. I wasn't aware of the cmd to exit ovn-controller. 
I think it helps ovn-controller to reconnect immediately, other than wait for 200s inactivity probe. Will add the cmd in in right after the encap IP update.

zshi-redhat commented:
> > @andreaskaris @zshi-redhat why dont we just exit ovn-controller if using tcp and ip changes via "ovn-appctl -t ovn-controller exit". ovnkube-node can execute this command.
> 
> @trozet sorry I missed this comment. I wasn't aware of the cmd to exit ovn-controller. I think it helps ovn-controller to reconnect immediately, other than wait for 200s inactivity probe. Will add the cmd right after the encap IP update.

added.

zshi-redhat commented:
the node-ip-migration e2e test failed, seems like ovn-encap-ip is not updated when node ip changed, investigating...

zshi-redhat commented:
> the node-ip-migration e2e test failed, seems like ovn-encap-ip is not updated when node ip changed, investigating...

It looks like `ovn-appctl -t ovn-controller exit` doesn't help in shortening the reconnect time.
It does restart the ovn-controller, but pod is not able to connect to other pods immediately after ovn-controller restarts.

After the first restart, the ovn-controller is observed to be restarted one more time during inactivity probe time and the following log (with updated encap IP) is captured in ovn-controller contianer after the second restart:
```
2023-01-16T04:26:48.086Z|00029|binding|INFO|rtoe-GR_ovn-worker: Claiming 02:42:ac:12:00:03 172.18.255.254/16
```
`172.18.255.254` is the new encap IP.  It seems the auto inactivity probe is needed for sb binding to happen?

trozet commented:
> > the node-ip-migration e2e test failed, seems like ovn-encap-ip is not updated when node ip changed, investigating...
> 
> It looks like `ovn-appctl -t ovn-controller exit` doesn't help in shortening the reconnect time. It does restart the ovn-controller, but pod is not able to connect to other pods immediately after ovn-controller restarts.
> 
> After the first restart, the ovn-controller is observed to be restarted one more time during inactivity probe time and the following log (with updated encap IP) is captured in ovn-controller contianer after the second restart:
> 
> ```
> 2023-01-16T04:26:48.086Z|00029|binding|INFO|rtoe-GR_ovn-worker: Claiming 02:42:ac:12:00:03 172.18.255.254/16
> ```
> 
> `172.18.255.254` is the new encap IP. It seems the auto inactivity probe is needed for sb binding to happen?

This seems unexpected to me. @dceara can you help here?

trozet commented:
I think we figured it out. There are a few things to do:
1. ovn-controller wasn't actually exiting, it was trying to update sbdb, which it couldnt do, before it would exit. To fix this add "--restart" to the ovn-controller exit cmd (you can see the code change on your server)
2. ovnkube-node was actually restarting ovn-controller when it would start because address manager had no primary addr. So the comparison would be true where:
```
[root@sdn-13 contrib]# kubectl logs -n ovn-kubernetes ovnkube-node-cdmbx -c ovnkube-node | egrep 'address manager primary addr:|ovn-controller exit'
I0118 19:29:28.981511    1461 node_ip_handler_linux.go:328] Node primary address: 172.18.0.2, address manager primary addr: <nil>
I0118 19:29:34.028051    1461 ovs.go:200] Exec(35): /usr/bin/ovn-appctl --timeout=5 -t ovn-controller exit --restart
I0118 19:29:34.324090    1461 node_ip_handler_linux.go:328] Node primary address: 172.18.0.2, address manager primary addr: 172.18.0.2
```
We need to avoid this extra restart.

3. The tests really run 4 tests. So that means 8 total restarts, cause on cleanup we change the ip back. This causes the container to go into crashloopbackoff which is why I think it was taking longer for the container to restart. The only thing I can think of is to "fix" it delete the ovnkube-node pod and let it respawn between tests, wait for it to go ready.

zshi-redhat commented:
> I think we figured it out. There are a few things to do:
> 
> 1. ovn-controller wasn't actually exiting, it was trying to update sbdb, which it couldnt do, before it would exit. To fix this add "--restart" to the ovn-controller exit cmd (you can see the code change on your server)

added `--restart` to ovn-controller exit cmd.

> 2. ovnkube-node was actually restarting ovn-controller when it would start because address manager had no primary addr. So the comparison would be true where:
> 
fixed to not trigger node primary address change when addressManager.nodePrimayAddr is nil.

> ```
> [root@sdn-13 contrib]# kubectl logs -n ovn-kubernetes ovnkube-node-cdmbx -c ovnkube-node | egrep 'address manager primary addr:|ovn-controller exit'
> I0118 19:29:28.981511    1461 node_ip_handler_linux.go:328] Node primary address: 172.18.0.2, address manager primary addr: <nil>
> I0118 19:29:34.028051    1461 ovs.go:200] Exec(35): /usr/bin/ovn-appctl --timeout=5 -t ovn-controller exit --restart
> I0118 19:29:34.324090    1461 node_ip_handler_linux.go:328] Node primary address: 172.18.0.2, address manager primary addr: 172.18.0.2
> ```
> 
> We need to avoid this extra restart.
> 
> 3. The tests really run 4 tests. So that means 8 total restarts, cause on cleanup we change the ip back. This causes the container to go into crashloopbackoff which is why I think it was taking longer for the container to restart. The only thing I can think of is to "fix" it delete the ovnkube-node pod and let it respawn between tests, wait for it to go ready.

added a validation to wait for ovn-encap-ip to be synced after rolling back to original node IP.

trozet commented:
"added a validation to wait for ovn-encap-ip to be synced after rolling back to original node IP."

I'm not sure how this exactly fixes the problem, but it looks like it did. My best guess is that the pod goes back to ready, so then crash loop backoff resets for next time when we kill the ovn-controller?

trozet reviewed (approved):
Thanks for the help on this one @zshi-redhat and @andreaskaris !

zshi-redhat commented:
> "added a validation to wait for ovn-encap-ip to be synced after rolling back to original node IP."
> 
> I'm not sure how this exactly fixes the problem, but it looks like it did. My best guess is that the pod goes back to ready, so then crash loop backoff resets for next time when we kill the ovn-controller?

Between two consecutive tests, the e2e tests revert the node ip change and don't wait enough time for the IP to be synced to ovn-encap-ip, so it takes relatively longer time in the next test for the ovnkube-node pod to settle down. The fix was to check the readiness of the external-ids:ovn-encap ip in a loop after each test run, then continue the next one: https://github.com/ovn-org/ovn-kubernetes/blob/master/test/e2e/node_ip_migration.go#L217

The same check was also added to each test run, after migrating the node IP: https://github.com/ovn-org/ovn-kubernetes/blob/master/test/e2e/node_ip_migration.go#L254. It is observed that it takes 2 or 3 loop for the encap-ip to be synced, this wait doesn't include the necessary time for ovn-controller to sync with sbdb after ovn-encap-ip is in place, which is why there is another 10 second sleep after ovn-encap-ip readiness check: https://github.com/ovn-org/ovn-kubernetes/blob/master/test/e2e/node_ip_migration.go#L223 .
